src/main/java/org/springframework/integration/aws/inbound/kinesis/KclMessageDrivenChannelAdapter.java

```diff
@@ -0,0 +1,400 @@
+import software.amazon.kinesis.exceptions.ThrottlingException;
+import software.amazon.kinesis.lifecycle.events.InitializationInput;
+import software.amazon.kinesis.lifecycle.events.LeaseLostInput;
+import software.amazon.kinesis.lifecycle.events.ProcessRecordsInput;
+import software.amazon.kinesis.lifecycle.events.ShardEndedInput;
+import software.amazon.kinesis.lifecycle.events.ShutdownRequestedInput;
+import software.amazon.kinesis.processor.RecordProcessorCheckpointer;
+import software.amazon.kinesis.processor.ShardRecordProcessor;
+import software.amazon.kinesis.processor.ShardRecordProcessorFactory;
+import software.amazon.kinesis.retrieval.KinesisClientRecord;
+
+/**
+ * The {@link MessageProducerSupport} implementation for receiving data from Amazon
+ * Kinesis stream(s) using AWS KCL.
+ *
+ * @author Hervé Fortin
+ *
+ * @since 2.1.0
+ */
+@ManagedResource
+@IntegrationManagedResource
+public class KclMessageDrivenChannelAdapter extends MessageProducerSupport implements DisposableBean {
+
+	private final String stream;
+
+	private final Set<KinesisShardOffset> shardOffsets = new HashSet<>();
+
+	private String consumerGroup = "SpringIntegration";
+
+	private InboundMessageMapper<byte[]> embeddedHeadersMapper;
+
+	private Scheduler scheduler;
+
+	private String region;
+
+	private InitialPositionInStreamExtended streamInitialSequence =
+			InitialPositionInStreamExtended.newInitialPosition(InitialPositionInStream.LATEST);
+
+	private int idleBetweenPolls;
+
+	private int consumerBackoff;
+
+	private long checkpointsInterval = 60_000L;
+
+	public KclMessageDrivenChannelAdapter(String streams) {
+		Assert.notNull(streams, "'streams' must not be null.");
+		this.stream = streams;
+	}
+
+	public void setConsumerGroup(String consumerGroup) {
+		Assert.hasText(consumerGroup, "'consumerGroup' must not be empty");
+		this.consumerGroup = consumerGroup;
+	}
+
+	/**
+	 * Specify an {@link InboundMessageMapper} to extract message headers embedded
+	 * into the record data.
+	 *
+	 * @param embeddedHeadersMapper the {@link InboundMessageMapper} to use.
+	 * @since 2.0
+	 */
+	public void setEmbeddedHeadersMapper(InboundMessageMapper<byte[]> embeddedHeadersMapper) {
+		this.embeddedHeadersMapper = embeddedHeadersMapper;
+	}
+
+	/**
+	 * Takes no action by default. Subclasses may override this if they need
+	 * lifecycle-managed behavior. Protected by 'lifecycleLock'.
+	 */
+	@Override
+	protected void doStart() {
+		super.doStart();
+
+		String workerId = UUID.randomUUID().toString();
+		RecordProcessorFactory recordProcessorFactory = new RecordProcessorFactory();
+
+		Region parsedRegion = (this.region == null) ? null : Region.of(this.region);
+		KinesisAsyncClient kinesisClient = KinesisAsyncClient.builder().region(parsedRegion).build();
+		CloudWatchAsyncClient cloudWatchClient = CloudWatchAsyncClient.builder().region(parsedRegion).build();
+		DynamoDbAsyncClient dynamoDBClient = DynamoDbAsyncClient.builder().region(parsedRegion).build();
```

Comment:
I would like to see all these clients injected into this component. I don't want to manage them internally. Although I might agree about a single region-based ctor and these client instantiations over there.